renderdoc/driver/vulkan/vk_core.cpp

```diff
@@ -705,6 +705,78 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
   Threading::SetTLSValue(debugMessageSinkTLSSlot, (void *)sink);
 }
 
+void WrappedVulkan::InsertPendingCommandBufferCallbacksEvent(VkCommandBuffer commandBuffer)
+{
+  // This occurs pre-baking as the event needs to be in the command buffer before vkEndCommandBuffer
+  // is called
+
+  VkResourceRecord *cmdRecord = GetRecord(commandBuffer);
+  VkPendingSubmissionCompleteCallbacks *pending =
+      cmdRecord->cmdInfo->pendingSubmissionCompleteCallbacks;
+  RDCASSERT(pending->event == VK_NULL_HANDLE);
+
+  if(pending->callbacks.empty())
+    return;
+
+  rdcarray<std::function<void()>> callbacks;
+  callbacks.swap(pending->callbacks);
+
+  const VkEventCreateInfo info = {VK_STRUCTURE_TYPE_EVENT_CREATE_INFO};
+  VkEvent event;
+  const VkResult vkr = ObjDisp(m_Device)->CreateEvent(Unwrap(m_Device), &info, NULL, &event);
+  CheckVkResult(vkr);
+
+  ObjDisp(commandBuffer)->CmdSetEvent(Unwrap(commandBuffer), event, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT);
+
+  pending->device = cmdRecord->cmdInfo->device;
+  pending->event = event;
+  pending->callbacks = std::move(callbacks);
+}
+
+void WrappedVulkan::AddPendingCommandBufferCallbacks(VkCommandBuffer commandBuffer)
+{
+  VkResourceRecord *cmdRecord = GetRecord(commandBuffer);
+  VkPendingSubmissionCompleteCallbacks *pending =
+      cmdRecord->bakedCommands->cmdInfo->pendingSubmissionCompleteCallbacks;
+
+  if(pending->callbacks.empty())
+    return;
+
+  RDCASSERT(pending->event != VK_NULL_HANDLE);
+
+  pending->AddRef();
+
+  SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
+  m_PendingCmdBufferCallbacks.push_back(pending);
+}
+
+void WrappedVulkan::CheckPendingCommandBufferCallbacks()
+{
+  SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
```

Comment:
This is fine right now and doesn't need to change, but it's worth having a mental pin in this that this function will hold the lock for quite a while depending on how long the `vkGetEventStatus` call takes, as well as the callbacks themselves.

If there's contention in the future this could be improved, but for now the only thing this would do is block one queue submission on one thread waiting to add to the array while another thread is submitting to a different queue and checking these.

Reply:
A trivial change could be to:

1. Acquire the lock
2. Move `m_PendingCmdBufferCallbacks` into a local
3. Release the lock
4. Do the checks/execution
5. Acquire the lock
6. _Prepend_ any remaining entries to `m_PendingCmdBufferCallbacks`
7. Release the lock

If not I can leave this as a comment as inspiration for some future person.